codelists/actions.py

```diff
@@ -491,30 +492,79 @@ def export_to_builder(*, version, author):
         CodeObj(version=draft, code=code_obj.code, status=code_obj.status)
         for code_obj in version.code_objs.all()
     )
+    # Note that a GET on the builder URL for a draft will update the draft again in
+    # case data has been updated since the draft was initially created.
+    # By default, we don't update it immediately in `export_to_builder` so as not to
+    # duplicate the update if the new draft is being passed on to the builder view
+    draft, _ = update_draft(draft=draft, source_version=version)
+    return draft
+
+
+def update_draft(*, draft, source_version=None):
+    """
+    Update this draft version by:
+    1) recreating the searches (either from a source version for a new draft, or
+       from the current version) in case data have been updated
+    2) checking for any new codes and ensuring the codeset is up to date
+    """
+    cached_updates = cache.get("updated_drafts", [])
+    new_codes = []
+    if draft.pk not in cached_updates:
+        _recreate_draft_searches(source_version or draft, draft)
+        new_codes = _check_new_concepts(draft)
+        cached_updates.append(draft.pk)
+        cache.set("updated_drafts", cached_updates, timeout=60 * 60)
+
+    return draft, new_codes
 
-    # Recreate each search.  This creates the SearchResults linked to the new draft.  We
-    # can't just copy them across, because the data may have been updated, and new
-    # matching concepts might have been imported.
-    #
-    # In future, we should be able to short-circuit this by keeping track of the release
-    # that version was created with.
-    for search in version.searches.all():
-        codes = do_search(version.coding_system, term=search.term, code=search.code)[
-            "all_codes"
-        ]
+
+def _recreate_draft_searches(source_version, draft):
+    """
+    Recreate each search.  This creates the SearchResults linked to the new/updated draft.  We
+    can't just copy them across, because the data may have been updated, and new
+    matching concepts might have been imported.
+    In future, we should be able to short-circuit this by keeping track of the release
+    that version was created with.
+    """
+    for search in source_version.searches.all():
+        codes = do_search(
+            source_version.coding_system, term=search.term, code=search.code
+        )["all_codes"]
         builder_actions.create_search(
             draft=draft, term=search.term, code=search.code, codes=codes
         )
+    cache_hierarchy(version=draft)
 
-    # This assert will fire if new matching concepts have been imported.  At the moment,
-    # the builder frontend cannot deal with a CodeObj with status ?  if any of its
-    # ancestors are included or excluded.  We will have to deal with this soon but for
-    # now fail loudly.
```

Comment:
This comment was a red herring; new matching concepts are actually fine.  The issue is when there are new concepts that don't match a search, but are a descendent of an included/excluded code, because the search recreation will not create `CodeObj`s on the draft for those codes